### ISSUE TITLE
Very imbalanced topic proportion?

### ISSUE THREAD

**Issue by zhimin-z** [MaartenGr/BERTopic#1423]
I completed the topic modeling as the tutorial recommend, but only find the topic seems very proportional. 
You see the first topic account for 70+% but the other topics account for very little in the proportion.

Here is my proportion:
![image](https://github.com/MaartenGr/BERTopic/assets/8592144/164ab4eb-634c-4314-b178-d720ade23754)

My hyperparameter tuning is as follows: 
```
import gensim.corpora as corpora
import pandas as pd
import wandb
import os

from gensim.parsing.preprocessing import strip_punctuation
from sklearn.feature_extraction.text import TfidfVectorizer
from gensim.models.coherencemodel import CoherenceModel
from bertopic.vectorizers import ClassTfidfTransformer
from sentence_transformers import SentenceTransformer
# from bertopic.representation import KeyBERTInspired
from bertopic import BERTopic
from hdbscan import HDBSCAN
from umap import UMAP

path_output = os.path.join(os.getcwd(), 'Result', 'RQ1', 'Special Topics')
path_model = os.path.join(os.getcwd(), 'Code', 'RQ1', 'Special Topic Modeling', 'Model')
if not os.path.exists(path_model):
    os.makedirs(path_model)

wandb_project = 'asset-management-topic-modeling'

os.environ["WANDB_API_KEY"] = XXXXXX
os.environ["TOKENIZERS_PARALLELISM"] = "true"
os.environ["WANDB__SERVICE_WAIT"] = "100"

# set default sweep configuration
config_defaults = {
    # Refer to https://www.sbert.net/docs/pretrained_models.html
    'model_name': 'all-mpnet-base-v2',
    'metric_distane': 'cosine',
    'calculate_probabilities': True,
    # 'reduce_frequent_words': True,
    'prediction_data': True,
    'low_memory': False,
    'random_state': 42,
    'ngram_range': 2,
}

config_sweep = {
    'method': 'grid',
    'metric': {
        'name': 'Coherence CV',
        'goal': 'maximize'
    },
    'parameters': {
        'n_components': {
            'values': [4,6,5]#[3, 4, 5, 6, 7],
        },
    }
}


class TopicModeling:
    def __init__(self, topic_type, min_cluster_size=20):
        # Initialize an empty list to store top models
        self.top_models = []
        self.path_model = path_model
        
        df = pd.read_json(os.path.join(path_output, 'labels.json'))
        if topic_type == 'anomaly':
            df = df[df['Challenge_type'] == 'anomaly']
            self.docs = df[df['Challenge_summary'] != 'na']['Challenge_summary'].tolist() + df[df['Challenge_root_cause'] != 'na']['Challenge_root_cause'].tolist()
        elif topic_type == 'solution':
            docs = df[df['Solution'] != 'na']['Solution'].tolist()
            self.docs = [strip_punctuation(doc) for doc in docs]
        
        config_defaults['min_cluster_size'] = min_cluster_size
        config_sweep['name'] = topic_type
        config_sweep['parameters']['min_samples'] = {
            'values': list(range(1, config_defaults['min_cluster_size'] + 1))
        }
        
    def __train(self):
        # Initialize a new wandb run
        with wandb.init() as run:
            # update any values not set by sweep
            run.config.setdefaults(config_defaults)

            # Step 1 - Extract embeddings
            embedding_model = SentenceTransformer(run.config.model_name)

            # Step 2 - Reduce dimensionality
            umap_model = UMAP(n_components=wandb.config.n_components, metric=run.config.metric_distane,
                              random_state=run.config.random_state, low_memory=run.config.low_memory)

            # Step 3 - Cluster reduced embeddings
            hdbscan_model = HDBSCAN(min_cluster_size=run.config.min_cluster_size,
                                    min_samples=wandb.config.min_samples, prediction_data=run.config.prediction_data)

            # Step 4 - Tokenize topics
            vectorizer_model = TfidfVectorizer(ngram_range=(1, run.config.ngram_range))

            # Step 5 - Create topic representation
            ctfidf_model = ClassTfidfTransformer()
            # ctfidf_model = ClassTfidfTransformer(reduce_frequent_words=run.config.reduce_frequent_words)

            # # Step 6 - Fine-tune topic representation
            # representation_model = KeyBERTInspired()

            # All steps together
            topic_model = BERTopic(
                embedding_model=embedding_model,
                umap_model=umap_model,
                hdbscan_model=hdbscan_model,
                vectorizer_model=vectorizer_model,
                ctfidf_model=ctfidf_model,
                # representation_model=representation_model,
                calculate_probabilities=run.config.calculate_probabilities
            )

            topic_model = topic_model.fit(self.docs)
            topic_model.reduce_topics(self.docs, nr_topics='auto')

            # Preprocess Documents
            documents = pd.DataFrame({"Document": self.docs,
                                      "ID": range(len(self.docs)),
                                      "Topic": topic_model.topics_})
            documents_per_topic = documents.groupby(
                ['Topic'], as_index=False).agg({'Document': ' '.join})
            cleaned_docs = topic_model._preprocess_text(
                documents_per_topic.Document.values)

            # Extract vectorizer and analyzer from BERTopic
            vectorizer = topic_model.vectorizer_model
            analyzer = vectorizer.build_analyzer()

            # Extract features for Topic Coherence evaluation
            tokens = [analyzer(doc) for doc in cleaned_docs]
            dictionary = corpora.Dictionary(tokens)
            corpus = [dictionary.doc2bow(token) for token in tokens]
            topic_words = [[words for words, _ in topic_model.get_topic(
                topic)] for topic in range(len(set(topic_model.topics_))-1)]

            coherence_cv = CoherenceModel(
                topics=topic_words,
                texts=tokens,
                corpus=corpus,
                dictionary=dictionary,
                coherence='c_v'
            )

            coherence_umass = CoherenceModel(
                topics=topic_words,
                texts=tokens,
                corpus=corpus,
                dictionary=dictionary,
                coherence='u_mass'
            )

            coherence_cuci = CoherenceModel(
                topics=topic_words,
                texts=tokens,
                corpus=corpus,
                dictionary=dictionary,
                coherence='c_uci'
            )

            coherence_cnpmi = CoherenceModel(
                topics=topic_words,
                texts=tokens,
                corpus=corpus,
                dictionary=dictionary,
                coherence='c_npmi'
            )

            wandb.log({'Coherence CV': coherence_cv.get_coherence()})
            wandb.log({'Coherence UMASS': coherence_umass.get_coherence()})
            wandb.log({'Coherence UCI': coherence_cuci.get_coherence()})
            wandb.log({'Coherence NPMI': coherence_cnpmi.get_coherence()})
            wandb.log({'Topic Number': topic_model.get_topic_info().shape[0] - 1})
            wandb.log({'Uncategorized Post Number': topic_model.get_topic_info().at[0, 'Count']})

            model_name = f'{config_sweep["name"]}_{run.id}'
            topic_model.save(os.path.join(self.path_model, model_name))

    def sweep(self):
        wandb.login()
        sweep_id = wandb.sweep(config_sweep, project=wandb_project)
        wandb.agent(sweep_id, function=self.__train)
```

The topics come from my best topic model with the highest coherence score, what can I do now...?
![image](https://github.com/MaartenGr/BERTopic/assets/8592144/d7c78736-8106-4718-9ea3-584ec20a8fe6)

**Comment by MaartenGr** [comment]
> The topics come from my best topic model with the highest coherence score.

To start off, topic coherence can be a very imprecise measure. It is merely an approximation of how coherent it might be but in practice, this also means that due to the subjective nature of coherence the metric is also imperfect. This means that purely following this metric often does not lead to the "best" topic model. "best" here is also quite subjective and should be defined by yourself depending on your use case. 

For example, the third topic model in your table might actually have a more balanced topic distribution, which is something you aim for. Moreover, it can happen that with more topics, the coherence slowly starts dropping. This is not necessarily an issue as a low number of topics might inflate the coherence score. 

In other words, do not only pursue coherence for selecting a topic model. Your third model might actually perform much better than the coherence score implies. 

> Is that possible to have a more balance topic clustering?

That depends on the underlying clustering algorithm. HDBSCAN, as it is a density-based algorithm, tends to find "unbalanced" clusters. Generally, we do not see a perfect balance of topics across documents since topics in documents are almost always unbalanced. What you can do is select a different algorithm that tends to create more "balance", such as k-Means. 

> Does reduce_frequent_words help? How about KeyBERTInspired?

These do not influence the cluster creation at all and for the purpose of "balance" are irrelevant.

> Is that due to `reduce_topics`? 

That might indeed happen. If you set `reduce_topics` to any value, whether that is `"auto"` or an integer, it might try to merge all those "micro-clusters" into a more general topic. This could indeed inflate and "unbalance" the topics. I would highly advise then not to use `nr_topics='auto'` nor `.reduce_topics`. 

> But I need this function for better clustering. 

This I do not necessarily agree with. Aside from the fact that "better clustering" is subjective, you can indirectly control the number of topics that are generated with the `min_cluster_size` parameter in HDBSCAN. The higher this value the fewer clusters are generated. The lower this value the more clusters are generated. Generally, I am more satisfied with the resulting topics using `min_cluster_size` compared to `nr_topics`. So I would advise skipping `nr_topics`. 

All in all, I think it is important to again mention that optimizing for an imperfect measure such as coherence will lead you to topic models that may not suit your use case. A human in the loop is incredibly important here. I can imagine not wanting the topic model with the lowest coherence score possible but going for the highest score will also lead to issues.

**Comment by zhimin-z** [comment]
> > The topics come from my best topic model with the highest coherence score.
> 
> To start off, topic coherence can be a very imprecise measure. It is merely an approximation of how coherent it might be but in practice, this also means that due to the subjective nature of coherence the metric is also imperfect. This means that purely following this metric often does not lead to the "best" topic model. "best" here is also quite subjective and should be defined by yourself depending on your use case.
> 
> For example, the third topic model in your table might actually have a more balanced topic distribution, which is something you aim for. Moreover, it can happen that with more topics, the coherence slowly starts dropping. This is not necessarily an issue as a low number of topics might inflate the coherence score.
> 
> In other words, do not only pursue coherence for selecting a topic model. Your third model might actually perform much better than the coherence score implies.
> 
> > Is that possible to have a more balance topic clustering?
> 
> That depends on the underlying clustering algorithm. HDBSCAN, as it is a density-based algorithm, tends to find "unbalanced" clusters. Generally, we do not see a perfect balance of topics across documents since topics in documents are almost always unbalanced. What you can do is select a different algorithm that tends to create more "balance", such as k-Means.
> 
> > Does reduce_frequent_words help? How about KeyBERTInspired?
> 
> These do not influence the cluster creation at all and for the purpose of "balance" are irrelevant.
> 
> > Is that due to `reduce_topics`?
> 
> That might indeed happen. If you set `reduce_topics` to any value, whether that is `"auto"` or an integer, it might try to merge all those "micro-clusters" into a more general topic. This could indeed inflate and "unbalance" the topics. I would highly advise then not to use `nr_topics='auto'` nor `.reduce_topics`.
> 
> > But I need this function for better clustering.
> 
> This I do not necessarily agree with. Aside from the fact that "better clustering" is subjective, you can indirectly control the number of topics that are generated with the `min_cluster_size` parameter in HDBSCAN. The higher this value the fewer clusters are generated. The lower this value the more clusters are generated. Generally, I am more satisfied with the resulting topics using `min_cluster_size` compared to `nr_topics`. So I would advise skipping `nr_topics`.
> 
> All in all, I think it is important to again mention that optimizing for an imperfect measure such as coherence will lead you to topic models that may not suit your use case. A human in the loop is incredibly important here. I can imagine not wanting the topic model with the lowest coherence score possible but going for the highest score will also lead to issues.

Hi, @MaartenGr Thanks for your suggestion. I tried those measures but there is another issue appearing. After removing `nr_topics` and `KeyBERTInspired`, the coherence score drops quite a lot (from 0.8 to 0.45 alike). As you said, the coherence score is not the most important metric to take care of, but this drop, after further validation experiment, is due to the removal of `KeyBERTInspired`. Namely, if I use `KeyBERTInspired`, the coherence score can jump a lot!!
If I remove it, the coherence score drops a lot!! I was confused about it, it is by design? or something wrong with my experiment setup?

**Comment by MaartenGr** [comment]
> Namely, if I use KeyBERTInspired, the coherence score can improve a lot!! If I remove it, the coherence score drops a lot!!

The method tends to favor the way topic coherence is calculated but since it optimizes for coherence, it does not suddenly create better topics. It is a method that I typically prefer using but it is not without its disadvantages. c-TF-IDF, for example, is much better at capturing domain-specific words. 

> I was confused about it, it is by design? or is something wrong with my experiment setup? The thing is, if I use KeyBERTInspired still, then duplicate topics persist and I cannot differentiate them by checking the representation words. 

I do not think there is something necessarily wrong with your experiment. Especially removing `nr_topics` is a helpful method but do note that it was not suggested to remove `KeyBERTInspired` as it merely acts as the way a topic is represented, not created. 

> Checking manually without taking care of any metrics among tons of models is painstaking and time-consuming, I hope to have one or two metrics to decide the best model quickly. 

I would not advise picking two metrics and only optimizing for those. What you can do instead is pick two metrics and take the top *n* topic models that were created and manually go through those to see whether they make sense. It would be especially helpful if the top *n* topic models were somewhat different from one another, for example, the number of topics they created. 

I think it is true in most cases that if you optimize for one or two metrics without manual inspection that you will miss out on the best model for your use case. 

> The thing is, if I use KeyBERTInspired still, then duplicate topics persist and I cannot differentiate them by checking the representation words. 

`KeyBERTInspired` does not influence the creation of topics in any way, it merely changes the way the created topics are represented. In other words, the names or keywords that they contain. You can still differentiate them if you use [multi-aspect representations](https://maartengr.github.io/BERTopic/getting_started/multiaspect/multiaspect.html), for example with ChatGPT or MMR, and/or if you check out the most representative documents within topics.

> If I do not use it, then the coherence score drops so much and this negatively impacts my experiment results in credibility.

I would be careful approaching it this way. A topic model is much much more than just topic coherence and diversity. I would highly advise reading through [Is automated topic model evaluation broken? the incoherence of coherence](https://proceedings.neurips.cc/paper/2021/file/0f83556a305d789b1d71815e8ea4f4b0-Paper.pdf). Evaluating a topic model is a complex task and just because an imperfect coherence score drops should definitely not be the sole reason for a large negative impact on credibility.

**Comment by zhimin-z** [comment]
Hi, @MaartenGr When doing the topic modeling, I noticed that if I decrease the `min_cluter_size`, then the topic number increases furiously, however, if I increase the `min_cluter_size`, the outlier number increases furiously. The former makes the topic interpretation extremely hard due to the fact that the topic representation is almost the same from human perspective, while the latter makes the `reduce_outlier` operation unideal because of the large number of outliers polluting the original topic representation. 

I have spent weeks over these tradeoffs and feel very frustrated, any practical suggestions?

**Comment by MaartenGr** [comment]
The things that you run into are a result of the underlying clustering algorithm, namely HDBSCAN. As you have seen, there are quite a number of things that you can optimize in order to reduce the issues that you face. 

Most notably, `.reduce_outliers` is still a good option I believe. The reason for this is the differentiation between *topic representation* and *topic assignment*. By itself `.reduce_outliers` only changes *topic assignment*, meaning that the outliers are now assigned to topics without actually influencing the representation, or words/phrases/ect., of those topics. If you want to then change the *topic representation* as a result of the *topic assignment*, you can run `.update_topics`. However, I do not believe that is always the right choice if it negatively influences the *topic representation*. Just by foregoing the `.update_topics` step, you can keep your high-quality topics and representations whilst still reducing outliers. 

The second thing that is of note here is that your struggles are with HDBSCAN specifically, so why not change it to something that fits better with your use case? In theory, BERTopic supports most clustering algorithms, so you could switch it over to k-Means, hierarchical clustering, sklearn's HDBSCAN, OPTICS, etc. Which works for you depends on the algorithm and your use case, so trying some different ones to get a feeling might help. 

>  while the latter makes the reduce_outlier operation unideal because of the large number of outliers polluting the original topic representation.

All in all, I think using `.reduce_outliers` without `.update_topics` would actually make the latter that you referred to still a valid solution. Just because a document is assigned to a topic does not mean that the topic representation should be influenced by that document.

**Comment by zhimin-z** [comment]
BYW, I found one thing: when the topic number is very large, let's say 170, topic term visualization does not show all the topics anymore. Currently, I can only see there are 72 topics in the generated figure. is that a limitation of this tool?

**Comment by zhimin-z** [comment]
> The things that you run into are a result of the underlying clustering algorithm, namely HDBSCAN. As you have seen, there are quite a number of things that you can optimize in order to reduce the issues that you face.
> 
> Most notably, `.reduce_outliers` is still a good option I believe. The reason for this is the differentiation between _topic representation_ and _topic assignment_. By itself `.reduce_outliers` only changes _topic assignment_, meaning that the outliers are now assigned to topics without actually influencing the representation, or words/phrases/ect., of those topics. If you want to then change the _topic representation_ as a result of the _topic assignment_, you can run `.update_topics`. However, I do not believe that is always the right choice if it negatively influences the _topic representation_. Just by foregoing the `.update_topics` step, you can keep your high-quality topics and representations whilst still reducing outliers.
> 
> The second thing that is of note here is that your struggles are with HDBSCAN specifically, so why not change it to something that fits better with your use case? In theory, BERTopic supports most clustering algorithms, so you could switch it over to k-Means, hierarchical clustering, sklearn's HDBSCAN, OPTICS, etc. Which works for you depends on the algorithm and your use case, so trying some different ones to get a feeling might help.
> 
> > while the latter makes the reduce_outlier operation unideal because of the large number of outliers polluting the original topic representation.
> 
> All in all, I think using `.reduce_outliers` without `.update_topics` would actually make the latter that you referred to still a valid solution. Just because a document is assigned to a topic does not mean that the topic representation should be influenced by that document.

I checked the assigned outliers, but they seldom make sense to be redistributed to the recognized topics. For optimizing outlier assignment, is that any other measures that I can tweak with?
